haystack/core/pipeline/base.py

```diff
@@ -1114,18 +1114,22 @@ def _write_component_outputs(
             if receiver_name not in inputs:
                 inputs[receiver_name] = {}
 
-            # If we have a non-variadic or a greedy variadic receiver socket, we can just overwrite any inputs
-            # that might already exist (to be reconsidered but mirrors current behavior).
-            if not is_socket_lazy_variadic(receiver_socket):
-                inputs[receiver_name][receiver_socket.name] = [{"sender": component_name, "value": value}]
-
-            # If the receiver socket is lazy variadic, and it already has an input, we need to append the new input.
-            # Lazy variadic sockets can collect multiple inputs.
+            if is_socket_lazy_variadic(receiver_socket):
+                _write_to_lazy_variadic_socket(
+                    inputs=inputs,
+                    receiver_name=receiver_name,
+                    receiver_socket_name=receiver_socket.name,
+                    component_name=component_name,
+                    value=value,
+                )
             else:
-                if not inputs[receiver_name].get(receiver_socket.name):
-                    inputs[receiver_name][receiver_socket.name] = []
-
-                inputs[receiver_name][receiver_socket.name].append({"sender": component_name, "value": value})
+                _write_to_standard_socket(
```

Comment:
Let's add a comment here too.
```suggestion
                # If the receiver socket is not lazy variadic, it is greedy variadic or non-variadic.
                # We overwrite with the new input if it's not _NO_OUTPUT_PRODUCED or if the current value is None.
                _write_to_standard_socket(
```